### PR TITLE
8319822: Use a linear-time algorithm for assert_different_registers()

### PR DESCRIPTION
**Pull request by theRealAph**
At the present time, `assert_different_registers()` uses an O(N**2) algorithm in assert_different_registers(). We can utilize RegSet to do it in O(N) time. This would be a useful optimization for all builds with assertions enabled.

In addition, it would be useful to be able to static_assert different registers. 

Also, I've taken the opportunity to expand the maximum size of a RegSet to 64 on 64-bit platforms.

I also fixed a bug: sometimes `noreg` is passed to `assert_different_registers()`, but it may only be passed once or a spurious assertion is triggered.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319822](https://bugs.openjdk.org/browse/JDK-8319822): Use a linear-time algorithm for assert_different_registers() (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16617/head:pull/16617` \
`$ git checkout pull/16617`

Update a local copy of the PR: \
`$ git checkout pull/16617` \
`$ git pull https://git.openjdk.org/jdk.git pull/16617/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16617`

View PR using the GUI difftool: \
`$ git pr show -t 16617`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16617.diff">https://git.openjdk.org/jdk/pull/16617.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16617#issuecomment-1805977396)

### CODE REVIEW

**Comment by TobiHartmann** [comment]
The build fails on Windows:
```
[2023-11-13T08:04:09,465Z] c:\sb\prod\1699862498\workspace\open\src\hotspot\cpu\x86\register_x86.hpp(402): error C2220: the following warning is treated as an error
[2023-11-13T08:04:09,465Z] c:\sb\prod\1699862498\workspace\open\src\hotspot\cpu\x86\register_x86.hpp(402): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
[2023-11-13T08:04:09,465Z] c:\sb\prod\1699862498\workspace\open\src\hotspot\cpu\x86\register_x86.hpp(415): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
[2023-11-13T08:04:09,465Z] lib/CompileGtest.gmk:94: recipe for target '/cygdrive/c/sb/prod/1699862498/workspace/build/windows-x64-open/hotspot/variant-server/libjvm/gtest/objs/static/BUILD_GTEST_LIBJVM_pch.obj' failed
[2023-11-13T08:04:09,465Z] make[3]: *** [/cygdrive/c/sb/prod/1699862498/workspace/build/windows-x64-open/hotspot/variant-server/libjvm/gtest/objs/static/BUILD_GTEST_LIBJVM_pch.obj] Error 1
```

**Comment by theRealAph** [comment]
> The build fails on Windows:

Argh! Thanks.

**Comment by stefank** [comment]
I started to review the patch and was wondering if this could be simplify to something like this?:
https://github.com/stefank/jdk/commit/f38c791793440b899ce6c4c9723470a5d4b18050

I tested this with this small section of temporary static_asserts:
https://github.com/stefank/jdk/commit/30da4d6abeee14e4e4f44034295f1bb0ad2e3016

Unfortunately, that didn't compile and I had make this change to get it to work:
https://github.com/stefank/jdk/commit/d6bda1a25e297865fd6b5da21184273d8825b922

**Comment by theRealAph** [comment]
> I started to review the patch and was wondering if this could be simplify to something like this?: [stefank@f38c791](https://github.com/stefank/jdk/commit/f38c791793440b899ce6c4c9723470a5d4b18050)

Sure, it could be done. This is a minor efficiency tweak.

**Comment by dean-long** [comment]
I ran into a case where I was doing assert_different_registers() on base() and index() from an Address.  How hard would it be to have assert_different_registers() support an Address as an argument?

**Comment by shqking** [comment]
> > The build fails on Windows:
> 
> Argh! Thanks.

I'm afraid the compiler warnings on Windows still exist. See the GHA test results.
Besides, the copyright year of `register.hpp` file should be updated to **2023**.

**Comment by theRealAph** [comment]
> I ran into a case where I was doing assert_different_registers() on base() and index() from an Address. How hard would it be to have assert_different_registers() support an Address as an argument?

Quite tricky, partly because `Address` isn't a shared type. I would have thought it best to make it explicit anyway, with something like
`different_registers(a.base(), a.index(), rsi, ...);`
We'd need to make sure that `base()` and `index()` return `noreg` whenever the `Address` is wrong, e.g. a literal.

**Comment by stefank** [comment]
> > I started to review the patch and was wondering if this could be simplify to something like this?: [stefank@f38c791](https://github.com/stefank/jdk/commit/f38c791793440b899ce6c4c9723470a5d4b18050)
> 
> Sure, it could be done. This is a minor efficiency tweak.

It tested the build performance before this PR, with the patch in this PR, and my simplified version. I can't see any performance difference on my MacBook M1. Is there any platform where this makes a bigger difference?

**Comment by stefank** [comment]
From the summary:
> In addition, it would be useful to be able to static_assert different registers.

As mentioned in https://github.com/openjdk/jdk/pull/16617#issuecomment-1807933886 this doesn't work unless we make the proposed small tweak. Do you want to make it in this PR, or should I propose that in a separate PR?

**Comment by theRealAph** [comment]
> > > I started to review the patch and was wondering if this could be simplify to something like this?: [stefank@f38c791](https://github.com/stefank/jdk/commit/f38c791793440b899ce6c4c9723470a5d4b18050)
> > 
> > Sure, it could be done. This is a minor efficiency tweak.
> 
> It tested the build performance before this PR, with the patch in this PR, and my simplified version. I can't see any performance difference on my MacBook M1. Is there any platform where this makes a bigger difference?
>
> Edit: I realize that since this doesn't always boil down to a constexpr, then the run time might be more interesting than the build time.

I don't think you'll be able to measure such a tiny difference in build performance, especially given all the other thiungs that are going on, and especially on a Great Big Out-Of_Order machine. But at runtime, the three-operand check can be performed with just three comparisons. I'm not fixated on this, though, and I can take the special cases out.

**Comment by theRealAph** [comment]
> Unfortunately, that didn't compile and I had make this change to get it to work: [stefank@d6bda1a](https://github.com/stefank/jdk/commit/d6bda1a25e297865fd6b5da21184273d8825b922)

Done, thanks and sorry for the delay.

File: src/hotspot/share/asm/register.hpp [hunk 1]
```diff
@@ -245,14 +253,44 @@ inline ReverseRegSetIterator<RegImpl> AbstractRegSet<RegImpl>::rbegin() {
 
 // Debugging support
 
+template<typename R, typename... Rx>
+inline constexpr bool different_registers(AbstractRegSet<R> allocated_regs, R first_register) {
+  return ! allocated_regs.contains(first_register);
```
Comment (eme64) [inline src/hotspot/share/asm/register.hpp:258]:
```suggestion
  return !allocated_regs.contains(first_register);
```

File: src/hotspot/share/asm/register.hpp [hunk 2]
```diff
@@ -245,14 +253,44 @@ inline ReverseRegSetIterator<RegImpl> AbstractRegSet<RegImpl>::rbegin() {
...
+}
+
+template<typename R, typename... Rx>
+inline constexpr bool different_registers(AbstractRegSet<R> allocated_regs, R first_register, Rx... more_registers) {
+  if (allocated_regs.contains(first_register))
+    return false;
```
Comment (eme64) [inline src/hotspot/share/asm/register.hpp:264]: Use curly scope brackets ;)

File: src/hotspot/share/asm/register.hpp [hunk 3]
```diff
@@ -245,14 +253,44 @@ inline ReverseRegSetIterator<RegImpl> AbstractRegSet<RegImpl>::rbegin() {
...
-  // Verify there are no equal entries.
-  for (size_t i = 0; i < ARRAY_SIZE(regs) - 1; ++i) {
-    for (size_t j = i + 1; j < ARRAY_SIZE(regs); ++j) {
-      assert(regs[i] != regs[j], "Multiple uses of register: %s", regs[i]->name());
+  if (! different_registers(first_register, more_registers...)) {
+    const R regs[] = { first_register, more_registers... };
+    // Find a duplicate entry.
+    for (size_t i = 0; i < ARRAY_SIZE(regs) - 1; ++i) {
+      for (size_t j = i + 1; j < ARRAY_SIZE(regs); ++j) {
+        assert(! regs[i]->is_valid() || regs[i] != regs[j],
```
Comment (eme64) [inline src/hotspot/share/asm/register.hpp:291]:
```suggestion
        assert(!regs[i]->is_valid() || regs[i] != regs[j],
```

File: src/hotspot/share/asm/register.hpp [hunk 4]
```diff
@@ -245,14 +253,44 @@ inline ReverseRegSetIterator<RegImpl> AbstractRegSet<RegImpl>::rbegin() {
...
+
 template<typename R, typename... Rx>
 inline void assert_different_registers(R first_register, Rx... more_registers) {
 #ifdef ASSERT
-  const R regs[] = { first_register, more_registers... };
-  // Verify there are no equal entries.
-  for (size_t i = 0; i < ARRAY_SIZE(regs) - 1; ++i) {
-    for (size_t j = i + 1; j < ARRAY_SIZE(regs); ++j) {
-      assert(regs[i] != regs[j], "Multiple uses of register: %s", regs[i]->name());
+  if (! different_registers(first_register, more_registers...)) {
```
Comment (eme64) [inline src/hotspot/share/asm/register.hpp:286]:
```suggestion
  if (!different_registers(first_register, more_registers...)) {
```

File: src/hotspot/share/asm/register.hpp [hunk 5]
```diff
@@ -245,14 +253,44 @@ inline ReverseRegSetIterator<RegImpl> AbstractRegSet<RegImpl>::rbegin() {
 
 // Debugging support
 
+template<typename R, typename... Rx>
+inline constexpr bool different_registers(AbstractRegSet<R> allocated_regs, R first_register) {
```
Comment (kimbarrett) [inline src/hotspot/share/asm/register.hpp:257]: "inline" is redundant with "constexpr".

File: src/hotspot/cpu/aarch64/register_aarch64.hpp [hunk 6]
```diff
@@ -69,8 +69,8 @@ class Register {
 
   constexpr Register() : _encoding(-1) {} // noreg
 
-  int operator==(const Register r) const { return _encoding == r._encoding; }
-  int operator!=(const Register r) const { return _encoding != r._encoding; }
+  constexpr bool operator==(const Register r) const { return _encoding == r._encoding; }
+  constexpr bool operator!=(const Register r) const { return _encoding != r._encoding; }
```
Comment (kimbarrett) [inline src/hotspot/cpu/aarch64/register_aarch64.hpp:73]: This seems unrelated to the rest of this change.  It also seems like something that should be done for all
of the register_<cpu> variants.

File: src/hotspot/share/asm/register.hpp [hunk 7]
```diff
@@ -245,14 +253,44 @@ inline ReverseRegSetIterator<RegImpl> AbstractRegSet<RegImpl>::rbegin() {
...
+    return false;
+  return different_registers(allocated_regs + first_register, more_registers...);
+}
+
+template<typename R, typename... Rx>
+inline constexpr bool different_registers(R first_register, Rx... more_registers) {
+  return different_registers(AbstractRegSet<R>(first_register), more_registers...);
+}
+
+template<typename R, typename... Rx>
```
Comment (kimbarrett) [inline src/hotspot/share/asm/register.hpp:273]: Rx is unused and not needed.  Similarly for 3-R overload.

File: src/hotspot/share/asm/register.hpp [hunk 8]
```diff
@@ -245,14 +253,44 @@ inline ReverseRegSetIterator<RegImpl> AbstractRegSet<RegImpl>::rbegin() {
...
+inline constexpr bool different_registers(R reg1, R reg2) {
+  return reg1 != reg2;
+}
+
+template<typename R, typename... Rx>
+inline constexpr bool different_registers(R reg1, R reg2, R reg3) {
+  return reg1 != reg2 && reg2 != reg3 && reg1 != reg3;
+}
```
Comment (kimbarrett) [inline src/hotspot/share/asm/register.hpp:281]: The 2-R and 3-R overloads are just an optimization (probably) of the variadic-R overload for small numbers
of arguments.  Given this is debug-only code, is it really worth the additional source code?

File: src/hotspot/share/asm/register.hpp (hunk 5 above)
Comment (kimbarrett) [inline src/hotspot/share/asm/register.hpp:257]: different_registers is only used by debug-only code in assert_different_registers.  Shouldn't all the overloads
for different_registers be within an `#ifdef ASSERT` block?

File: src/hotspot/share/asm/register.hpp [hunk 9]
```diff
@@ -93,67 +93,75 @@ template <class RegImpl> class ReverseRegSetIterator;
...
     return AbstractRegSet(bits);
   }
 
-  uint size() const { return population_count(_bitset); }
+  constexpr bool contains(RegImpl reg) {
+    return (AbstractRegSet(reg).bits() & bits()) != 0;
+  }
 
-  uint32_t bits() const { return _bitset; }
+  constexpr uint size() const { return population_count(_bitset); }
```
Comment (kimbarrett) [inline src/hotspot/share/asm/register.hpp:163]: population_count is currently not constexpr.  I'm surprised this doesn't lead to warnings for guaranteed
non-constexpr body of a constexpr function.  I'm pretty sure I've seen such warnings from some compiler.
Note that I think there's no reason for it to not be constexpr, other than preceding our use of C++11/14 and
not yet updated accordingly.  https://bugs.openjdk.org/browse/JDK-8323952

**Review by kimbarrett — CHANGES_REQUESTED** [review]

File: src/hotspot/share/asm/register.hpp (hunk 5 above)
Comment (theRealAph) [reply src/hotspot/share/asm/register.hpp:257]: I could do so, but that would lose the ability to do `static_assert(different_registers(...`. I don't think that  `static_assert` depends on `ASSERT`. I'm happy to make this patch debug-only, though, if you prefer.

**Comment by theRealAph** [comment]
> I started to review the patch and was wondering if this could be simplify to something like this?: [stefank@f38c791](https://github.com/stefank/jdk/commit/f38c791793440b899ce6c4c9723470a5d4b18050)
> 
> I tested this with this small section of temporary static_asserts: [stefank@30da4d6](https://github.com/stefank/jdk/commit/30da4d6abeee14e4e4f44034295f1bb0ad2e3016)
> 
> Unfortunately, that didn't compile and I had make this change to get it to work: [stefank@d6bda1a](https://github.com/stefank/jdk/commit/d6bda1a25e297865fd6b5da21184273d8825b922)

OK, so I'm not going to do that, then.

File: src/hotspot/cpu/aarch64/register_aarch64.hpp (hunk 6 above)
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/register_aarch64.hpp]: It was related to another reviewer's comments, but we don't need it

File: src/hotspot/share/asm/register.hpp [hunk 10]
```diff
@@ -245,14 +253,35 @@ inline ReverseRegSetIterator<RegImpl> AbstractRegSet<RegImpl>::rbegin() {
...
 // Debugging support
 
+template<typename R, typename... Rx>
+inline constexpr bool different_registers(AbstractRegSet<R> allocated_regs, R first_register) {
+  return !allocated_regs.contains(first_register);
+}
+
+template<typename R, typename... Rx>
+inline constexpr bool different_registers(AbstractRegSet<R> allocated_regs, R first_register, Rx... more_registers) {
+  if (allocated_regs.contains(first_register)) {
```
Comment (stefank) [inline src/hotspot/share/asm/register.hpp:263]: FWIW, while first reading this I was looking for the base case of the recursion (the previous versions had some extra specializations). To me it looks like the base case is written in both this function and the function above. I would prefer to have the implementation inside one function only and change this function to use:
```
  if (!different_registers(allocated_regs, first_register)) {
```
I think this could make it a bit clearer, but if you prefer the current style, I think that's fine as well.

**Review by stefank — APPROVED** [review]
Approved. I've written some suggestions that I would prefer, but that are not strictly necessary before integration.

File: src/hotspot/share/asm/register.hpp (hunk 10 above)
Comment (theRealAph) [reply src/hotspot/share/asm/register.hpp:263]: I'd prefer to stick with what I have, because it's a bit more direct and slightly simpler runtime code.

**Comment by theRealAph** [comment]
> > > I started to review the patch and was wondering if this could be simplify to something like this?: [stefank@f38c791](https://github.com/stefank/jdk/commit/f38c791793440b899ce6c4c9723470a5d4b18050)
> > 
> > 
> > Sure, it could be done. This is a minor efficiency tweak.
> 
> It tested the build performance before this PR, with the patch in this PR, and my simplified version. I can't see any performance difference on my MacBook M1. Is there any platform where this makes a bigger difference?
> 
> Edit: I realize that since this doesn't always boil down to a constexpr, then the run time might be more interesting than the build time.

You won't see very much, if any, because other things dominate. The main advantage, going forward, is that much of this can be constexpr'd, once I find out how to test on Windows.

**Review by kimbarrett — CHANGES_REQUESTED** [review]

File: src/hotspot/share/asm/register.hpp (hunk 7 above)
Comment (theRealAph) [reply src/hotspot/share/asm/register.hpp:274]: Isn't it? It seems to me to be used for the next line.

**Comment by theRealAph** [comment]
> Just a code-style review.
> 
> Question: could there be some sort of regression test for this, with different examples and edge cases?

I have no idea, really. assert_different_registers is used all over the place, and I'm going for bootcycle and tier1.

File: src/hotspot/share/asm/register.hpp (hunk 7 above)
Comment (kimbarrett) [reply src/hotspot/share/asm/register.hpp:274]: I think that comment about unused Rx was on the overload `different_registers(ARS<R>, R)`, which
had an Rx parameter in an old version of this change.  It's okay now.  And there's no longer any 3-R
overload.

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Comment by tstuefe** [comment]
> > Just a code-style review.
> > Question: could there be some sort of regression test for this, with different examples and edge cases?
> 
> I have no idea, really. assert_different_registers is used all over the place, and I'm going for bootcycle and tier1.

You could write a death test gtest. Like this:

```
TEST_VM_ASSERT_MSG(AssemblerAArch64, assert_different_regs, ".*Multiple uses of register: c_rarg0.*") {
  Register reg1 = r0;
  Register reg2 = r0;
  assert_different_registers(reg1, reg2);
}
```

**Comment by theRealAph** [comment]
> > > Just a code-style review.
> > > Question: could there be some sort of regression test for this, with different examples and edge cases?
> > 
> > 
> > I have no idea, really. assert_different_registers is used all over the place, and I'm going for bootcycle and tier1.
> 
> You could write a death test gtest. Like this:
> 
> ```
> TEST_VM_ASSERT_MSG(AssemblerAArch64, assert_different_regs, ".*Multiple uses of register: c_rarg0.*") {
>   Register reg1 = r0;
>   Register reg2 = r0;
>   assert_different_registers(reg1, reg2);
> }
> ```

I could, but I don't think there's much point. `assert_different_registers()` is used so much that it'll get thoroughly tested in the positive cases, at least. Do you think this is important?

File: src/hotspot/share/asm/register.hpp [hunk 11]
```diff
@@ -93,67 +93,75 @@ template <class RegImpl> class ReverseRegSetIterator;
 // A set of registers
 template <class RegImpl>
 class AbstractRegSet {
-  uint32_t _bitset;
+  size_t _bitset;
```
Comment (tstuefe) [inline src/hotspot/share/asm/register.hpp:96]: Why couple the number of possible registers to the memory size? Why not uint64_t?

Comment (tstuefe) [reply src/hotspot/share/asm/register.hpp:96]: 64-bit makes sense.

I think this had been broken for ppc where the number of vector registers can exceed 32 (https://github.com/openjdk/jdk/blob/91101f0d4fc8e06d0d74e06361db6ac87efeeb8e/src/hotspot/cpu/ppc/register_ppc.hpp#L378)

Unless I am mistaken, `assert_different_registers`, if applied to VSR32 and up, would never have fired.

File: src/hotspot/share/asm/register.hpp [hunk 12]
```diff
@@ -93,67 +93,75 @@ template <class RegImpl> class ReverseRegSetIterator;
...
+  }
 
 public:
 
-  AbstractRegSet() : _bitset(0) { }
+  constexpr AbstractRegSet() : _bitset(0) { }
 
-  AbstractRegSet(RegImpl r1) : _bitset(1 << r1->encoding()) { }
+  constexpr AbstractRegSet(RegImpl r1)
+    : _bitset(r1->is_valid() ? size_t(1) << r1->encoding() : 0) { }
```
Comment (tstuefe) [inline src/hotspot/share/asm/register.hpp:109]: If my assumption from above is correct, we never noticed ppc being broken since _bitset would have been 0 if encoding > 31.

Could you add something like:

```
STATIC_ASSERT(RegImpl::number_of_registers <= 64);
```

?

**Review by tstuefe — COMMENTED** [review]
> I could, but I don't think there's much point. assert_different_registers() is used so much that it'll get thoroughly tested in the positive cases, at least. Do you think this is important?

See my remarks. I am mainly concerned about exceeding the range for the bit set. If you add the proposed static assert, I think we are good.

File: src/hotspot/share/asm/register.hpp (hunk 11 above)
Comment (theRealAph) [reply src/hotspot/share/asm/register.hpp:96]: OK, I get it. I was sort-of thinking that on the 32-bit platforms we support we don't ever have more than 32 registers in a set, but maybe that's not true. I certainly don't want to slow down 32-bit platforms by burdening them with double-word operations for something that can never happen.

File: src/hotspot/share/asm/register.hpp (hunk 12 above)
Comment (theRealAph) [reply src/hotspot/share/asm/register.hpp:109]: OK.

File: src/hotspot/share/asm/register.hpp (hunk 11 above)
Comment (tstuefe) [reply src/hotspot/share/asm/register.hpp:96]: Okay. Well, if we have a static assert, we will notice if we have more registers than fit the bitset.

Comment (theRealAph) [reply src/hotspot/share/asm/register.hpp]: So I'm getting this bizarre failure on arm32. I'm guessing it's actually a compiler bug, but I suppose it might be some dusty corner of C++ to do with template arg substitution. Any thoughts?

```
In file included from /home/runner/work/jdk/jdk/src/hotspot/share/utilities/globalDefinitions.hpp:29,
                 from /home/runner/work/jdk/jdk/src/hotspot/share/nmt/memflags.hpp:28,
                 from /home/runner/work/jdk/jdk/src/hotspot/share/memory/allocation.hpp:29,
                 from ad_arm.hpp:30,
                 from ad_arm.cpp:28:
/home/runner/work/jdk/jdk/src/hotspot/share/asm/register.hpp: In instantiation of ‘class AbstractRegSet<RegisterImpl*>’:
/home/runner/work/jdk/jdk/src/hotspot/share/asm/register.hpp:272:30:   required from ‘constexpr bool different_registers(R, Rx ...) [with R = RegisterImpl*; Rx = {RegisterImpl*, RegisterImpl*}]’
/home/runner/work/jdk/jdk/src/hotspot/share/asm/register.hpp:278:27:   required from ‘void assert_different_registers(R, Rx ...) [with R = RegisterImpl*; Rx = {RegisterImpl*, RegisterImpl*}]’
/home/runner/work/jdk/jdk/src/hotspot/cpu/arm/arm.ad:8984:52:   required from here
/home/runner/work/jdk/jdk/src/hotspot/share/asm/register.hpp:96:26: error: ‘number_of_registers’ is not a member of ‘RegisterImpl*’
   96 |   STATIC_ASSERT(RegImpl::number_of_registers <= 64);
      |                          ^~~~~~~~~~~~~~~~~~~
/home/runner/work/jdk/jdk/src/hotspot/share/utilities/debug.hpp:287:44: note: in definition of macro ‘STATIC_ASSERT’
  287 | #define STATIC_ASSERT(Cond) static_assert((Cond), #Cond)
      |                                            ^~~~
```

Comment (theRealAph) [reply src/hotspot/share/asm/register.hpp]: I just changed the static assert to a regular assert, in the hope that it might work better on arm32.

Comment (tstuefe) [reply src/hotspot/share/asm/register.hpp]: Urgh. 

Seems on arm32 (and on zero, but I guess that does not matter) Register is a typedef to a pointer:

https://github.com/openjdk/jdk/blob/9686e804a2b058955ff88149c54a0a7896c0a2eb/src/hotspot/cpu/arm/register_arm.hpp#L136

and RegImpl is a pointer type, so it cannot have members.

On all other platforms, Register is a class.

I don't see a quick fix other than to make the assert dependent on !arm !zero. Or, leave it for a followup fix.

---

BTW, note that on Arm32, we can have 64 float registers:

https://github.com/openjdk/jdk/blob/9686e804a2b058955ff88149c54a0a7896c0a2eb/src/hotspot/cpu/arm/register_arm.hpp#L178

So, I would make the underlying type 64-bit for 32-bit builds too.

Comment (tstuefe) [reply src/hotspot/share/asm/register.hpp]: BTW Would the runtime assert not prevent this from used as constexpr?

Comment (theRealAph) [reply src/hotspot/share/asm/register.hpp]: It doesn't, and I just saw the problem: on Arm, `Register` is a pointer type, whereas on other ports it's a class type.

Comment (theRealAph) [reply src/hotspot/share/asm/register.hpp]: Looks like our comments crossed over. I'm building Zero and it's OK so far.

Comment (theRealAph) [reply src/hotspot/share/asm/register.hpp]: I guess Arm never got converted. That port needs some love.

Comment (tstuefe) [reply src/hotspot/share/asm/register.hpp]: Or, deprecation.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 13]
```diff
@@ -68,6 +68,8 @@
 
 #include <sys/types.h>
 
+static_assert(different_registers(zr, sp), "fucked");
```
Comment (tstuefe) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:71]: Debugging remnant?

**Review by tstuefe — APPROVED** [review]
Looks good to me. Thanks for working in my remarks.

**Review by kimbarrett — APPROVED** [review]
Still looks good.
